liquidhaskell-boot/src/Language/Haskell/Liquid/Constraint/RewriteCase.hs

```diff
@@ -0,0 +1,144 @@
+import           Language.Haskell.Liquid.Types.RType
+
+import           Data.Maybe
+import           Data.Tuple
+
+import qualified Data.HashMap.Strict as M
+import qualified Data.HashSet        as S
+
+getCaseRewrites :: CGEnv -> SpecType -> LocalRewrites
+getCaseRewrites γ spec =
+    let Reft (_, refinement) = ur_reft $ rt_reft spec
+        -- Names of constustors both type constructors and data constructors
+        ctors   = toSet $ seBinds  $ constEnv γ
+        -- All the global names (top level functions, types, etc...)
+        globals = toSet $ reGlobal $ renv     γ
+    in unloop
+       $ concatMap (uncurry $ unify ctors globals)
+       $ groupUnifiableEqualities 
+       $ getEqualities refinement
+    where toSet = S.fromList . M.keys
+
+-- | Generates substitutions for non-global variables that make @e1@ and @e2@
+-- equal.
+--
+-- If @v@ is not global, and @C@ is a data constructor
+--
+--  * @v@ and @e2@ produces @(v, e2)@
+--  * @e1@ and @v@ produces @(e1, v)@
+--  * @C a₁ ... aₙ@ and @C b₁ ... bₙ@ produces the substitutions from unifying
+--    @(a₁, b₁), ..., (aₙ, bₙ)@
+--
+-- If any unification fails, the substitutions from the unifications that
+-- succeed are still produced.
+--
+unify :: S.HashSet Symbol -> S.HashSet Symbol -> Expr -> Expr -> [(Symbol, Expr)]
+unify ctors globals = go
+    where
+        go e1 e2 | e1 == e2 = []
+        -- NOTE: We don't need to check for ECst because the expressions arent
+        -- elaborated
+        go (EVar s1) e2 | isLocal s1 = [(s1, e2)]
+        go e1 (EVar s2) | isLocal s2 = [(s2, e1)]
+        -- TODO: Tecnically we could also unify under lambdas but you have to be
+        -- carefull about alpha equivalence idk if the effort is worth it.
+        go e1 e2 
+            -- Performing the unification under constructor is safe because 
+            -- C a₁ ... aₙ = C b₁ ... bₙ ⟺ ∀ n . a₁ = bₙ
+            | (EVar name1 , args1) <- splitEApp e2
+            , (EVar name2 , args2) <- splitEApp e1
+            , name1 == name2
+            , isCtor name1
+            , length args1 == length args2
+            = concat $ zipWith go args1 args2
+        go _ _ = []
+
+        isCtor  name = name `S.member` ctors
+        isLocal name = not (name `S.member` globals 
+                           || name `S.member` ctors
+                           || isPrefixOfSym anfPrefix name)
+
+
+-- | Given a list of equalities this function produces the equalities that
+-- result from applying transitivity exactly once through terms that are not
+-- unifiable. For instance, if we have @[e1=e2, e2=e3, e1=e4]@ this function
+-- will produce @[e1=e3, e2=e4]@.
+--
+-- Some equalities are not produced if more than two equalities refer to the
+-- same expression. For instance, @[e1=e2, e1=e3, e1=e4]@, becase @e1@ appears
+-- three times, the equality @e3=e4@ won't be produced. This is an heuristic
+-- to avoid producing equalities that can later produce redundant rewrites.
```

Comment:
Sorry I don't understand you mean `[(Expr, [Expr])]`?

Reply:
I think the following code would honor the description exactly.
```diff
 -- | Given a list of equalities this function produces the equalities that
 -- result from applying transitivity exactly once. For instance, if we have
 -- @[e1=e2, e2=e3, e1=e4]@ this function will produce @[e1=e3, e2=e4]@.
---                                                                                                                  
--- Some equalities are not produced if more than two equalities refer to the
--- same expression. For instance, @[e1=e2, e1=e3, e1=e4]@, becase @e1@ appears
--- three times, the equality @e3=e4@ won't be produced. This is an heuristic
--- to avoid producing equalities that can later produce redundant rewrites.
 groupUnifiableEqualities :: [(Expr, Expr)] -> [(Expr, Expr)]
 groupUnifiableEqualities = concatMap (concat . mkEqs) . grouping
   where                                                                                                                                                                                       
-    mkEqs (e1:es) = [ (e1, e) | e <- es ]
+    mkEqs (e1:es) = [ (e1, e) | e <- es ] : mkEqs es
     mkEqs _       = []

     grouping eqs = fmap snd $ M.groupList $ eqs ++ fmap swap eqs
```